### PR TITLE
Migration to DD4Hep for RPC db Loader

### PR DESCRIPTION
**Pull request by slomeo**
**PR description:**

This PR (only for RPC) followed what I did for CSC in #32001  . 

**PR validation:**

1) validation by "cmsRun CondTools/Geometry/test/rpcgeometrywriter.py" (with DD4hep flag set False or True):

Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 11-Nov-2020 13:26:09.163 CET

MessageLogger Summary

Severity    # Occurrences   Total Occurrences
--------    -------------   ---------------

dropped waiting message count 0

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

 2) validation by printouts for both DD & DD4Hep build methods within the RPCGeometryParsFromDD Class in order to compare some DetId parameters (for example the 637566989 detid):

//DD
1, RPCGeometryParsFromDD, detid: 637566989 name: MB1RPC_IGasLeft number of Strips: 90
2, RPCGeometryParsFromDD, dpar.size() == 3, width: 103 length: 58.45 thickness: 0.2
4, RPCGeometryParsFromDD, tran.x(): 413.675 tran.y(): 37.6 tran.z(): -470.45
5, RPCGeometryParsFromDD, x.X(), x.Y(), x.Z(), y.X(), y.Y(), y.Z(), z.X(), z.Y(), z.Z():
5, RPCGeometryParsFromDD 3.06162e-16, -1, 3.12322e-16, -4.78895e-16, -8.5632e-16, 1, -1, 6.40079e-16, 1.15045e-17

//DD4Hep
1, RPCGeometryParsFromDD, detid: 637566989 name: MB1RPC_IGasLeft number of Strips: 90
2, RPCGeometryParsFromDD, dd4hep::Box, width: 103 length: 58.45 thickness: 0.2
4, RPCGeometryParsFromDD, tran.x(): 413.675 tran.y(): 37.6 tran.z(): -470.45
5, RPCGeometryParsFromDD, x.X(), x.Y(), x.Z(), y.X(), y.Y(), y.Z(), z.X(), z.Y(), z.Z():
5, RPCGeometryParsFromDD 2.13888e-16, -1, 2.00422e-16, -9.9934e-17, -4.99491e-16, 1, -1, 3.02876e-16, -1.22527e-16

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

3) validation by "runTheMatrix.py -l 25202.1" (because I edited the RPCGeometryParsFromDD class and not only the DB Class):

25202.1_TTbar_13+TTbar_13+DIGIUP15APVSimu_PU25+RECOUP15_PU25+HARVESTUP15_PU25 Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED  - time date Wed Nov 11 13:59:46 2020-date Wed Nov 11 13:32:32 2020; exit: 0 0 0 0
1 1 1 1 tests passed, 0 0 0 0 failed

**if this PR is a backport please specify the original PR and why you need to backport that PR:**

nothing special

### CODE REVIEW

**Comment by slomeo** [comment]
@cvuosalo : please take a look at https://cmssdt.cern.ch/lxr/search?%21v=CMSSW_11_2_X_2020-11-10-2300&_filestring=&_string=RPCRecoIdealDBLoader 

Do I have to edit all these .py files adding 

_process.RPCGeometryWriter = cms.EDAnalyzer("RPCRecoIdealDBLoader",
                                           fromDD4Hep = cms.untracked.bool(False))_

instead of

_process.RPCGeometryWriter = cms.EDAnalyzer("RPCRecoIdealDBLoader")_ ?

**Comment by civanch** [comment]
please test

File: Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc [hunk 1]
```diff
@@ -1,31 +1,38 @@
...
 #include "DataFormats/GeometryVector/interface/Basic3DVector.h"
-
 #include "CLHEP/Units/GlobalSystemOfUnits.h"
-
 #include <iostream>
 #include <algorithm>
+#include <DetectorDescription/DDCMS/interface/DDFilteredView.h>
+#include <DetectorDescription/DDCMS/interface/DDCompactView.h>
+#include "DetectorDescription/DDCMS/interface/DDSpecParRegistry.h"
+#include "DataFormats/Math/interface/CMSUnits.h"
```
Comment (cvuosalo) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc:25]: Why is `CMSUnits.h` included? I don't see anything from it used in this file.

File: CondTools/Geometry/test/rpcgeometrywriter.py [hunk 2]
```diff
@@ -4,6 +4,8 @@
 process.load('CondCore.CondDB.CondDB_cfi')
 process.load('Configuration.StandardSequences.GeometryExtended_cff')
 process.load('Geometry.MuonNumbering.muonNumberingInitialization_cfi')
+process.load("Geometry.MuonNumbering.muonGeometryConstants_cff")
+process.load('Configuration.StandardSequences.DD4hep_GeometrySim_cff')
```
Comment (cvuosalo) [inline CondTools/Geometry/test/rpcgeometrywriter.py:8]: I don't think it is right to include this DD4hep geometry. This file needs to work with both old DD and DD4hep.

File: Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc [hunk 3]
```diff
@@ -75,18 +102,84 @@ void RPCGeometryParsFromDD::buildGeometry(DDFilteredView& fview,
...
+      const double dpar4 = geant_units::operators::convertMmToCm(dpar[4]);
+      const double dpar8 = geant_units::operators::convertMmToCm(dpar[8]);
+      const double dpar0 = geant_units::operators::convertMmToCm(dpar[0]);
+      pars = {dpar4, dpar8, dpar0, 0.4, numbOfStrips.doubles()[0]};
+      edm::LogVerbatim("RPCGeometryParsFromDD")
+          << " (3), else, dpar[4]: " << dpar4 << " dpar[8]: " << dpar8 << " dpar[0]: " << dpar0;
     }
-
-    const std::vector<double> vtra = {tran.x(), tran.y(), tran.z()};
+    const std::vector<double> vtra = {geant_units::operators::convertMmToCm(tran.x()),
```
Comment (cvuosalo) [inline Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc:119]: This change is worrisome. The old DD code should not need to be changed. Why would the units convention change now for old DD?

File: CondTools/Geometry/test/rpcgeometrywriter.py (hunk 2 above)
Comment (slomeo) [reply CondTools/Geometry/test/rpcgeometrywriter.py:8]: @cvuosalo : I added the same line for PR #32001 (for CSC). Without it an error appears:

----- Begin Fatal Exception 12-Nov-2020 09:04:49 CET-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing global begin Run run: 1
   [1] Calling method for module RPCRecoIdealDBLoader/'RPCGeometryWriter'
Exception Message:
No data of type "DDCompactView" with label "" in record "IdealGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
%MSG-i DDCompactViewImpl:  12-Nov-2020 09:04:49 CET EndJob
DDD transient representation has been destructed.
%MSG

File: Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc (hunk 1 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc:25]: @cvuosalo : I deleted this #include

File: Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc (hunk 3 above)
Comment (slomeo) [reply Geometry/RPCGeometryBuilder/src/RPCGeometryParsFromDD.cc:119]: @cvuosalo : the same code is present in the RPCGeometryBuilder Class. Without it in the Log file I found:

RPCGeometryParsFromDD, tran.x(): 4136.75 tran.y(): 376 tran.z(): -4704.5 // for DD
RPCGeometryParsFromDD, tran.x(): 413.675 tran.y(): 37.6 tran.z(): -470.45 // for DD4Hep

so I think that's geant4_units are necessary

**Comment by civanch** [comment]
please test

**Comment by cvuosalo** [comment]
@slomeo We have to be very careful with units in this code. For old DD, when the geometry is written and then read from the DB, it goes through these steps:
1. Read values from XML as mm.
2. Build geometry and store in DB as mm.
3. Read geometry from DB and convert mm to cm.

For DD4hep, this process has to  produce the same results. You need to make the DD4hep method consistent with `Geometry/RPCGeometryBuilder/src/RPCGeometryBuilderFromCondDB.cc`. You might want to change this file, but you have to make sure the three-step process ends up with the right units. 

You should also check that your changes to the CSC DB loader handle the units correctly.

**Comment by cvuosalo** [comment]
@slomeo To follow up on my last comment, for backward compatibility, we need to keep the units as they were. As in the past, mm need to be stored in the DB. That means when DD4hep values are read in the DB loader, they need to be converted to mm.

**Comment by slomeo** [comment]
@cvuosalo : ok, if I understood well in order to reproduce the same results for DD & DD4HEP inside the RPCGeometryParsFromDD, I have to restore "mm" for DD (i.e no geant4_units) and set to "mm" the values I got from DD4Hep  (i.e putting /mm or x10). This is why in RPCGeometryBuilderFromCondDB.cc there are lines such as (for example) "Surface::PositionType pos(*(tranStart) / cm, *(tranStart + 1) / cm, *(tranStart + 2) / cm);" where "/cm" is present.  
Is it so? 
I'll try to perform this change and I'll check it with myLog.log

**Comment by ianna** [comment]
@cvuosalo - as discussed at the DD4hep meeting and suggested by the DD4hep developers using the dd4hep units in the dd4hep-related code would spare us from having yet another code migration when they'll switch to the G4 units internally.

**Comment by slomeo** [comment]
@ianna @cvuosalo I'll wait to edit (again) the code of this PR till this afternoon chat (@civanch at 17:30 ?)

**Comment by cvuosalo** [comment]
In the DB loader, the values obtained from DD4hep need to be converted to mm. The units currently returned by DD4hep are cm; in the future, they will be switched to mm. The code in the DB loader can be written so that this upcoming change in DD4hep will not require a further change in the DB loader code. For example, line 156 of `RPCGeometryParsFromDD.cc` gets a translation length from DD4hep:
```
tran[0] = static_cast<double>(fview.translation().X());
```
The translation value needs to be converted to mm. A new function should be created to do the conversion, so this line would become:
```
tran[0] = convertDD4hepUnitsToMm(static_cast<double>(fview.translation().X()));
```
This function could be added to `DetectorDescription/DDCMS/interface` and would look like this:
```
template <class NumType>
inline constexpr NumType convertDD4hepUnitsToMm(NumType length)  // DD4hep length -> millimeters
{
  return (length / dd4hep::mm);
}
```

**Comment by ianna** [comment]
I'd stick with a [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and simply use `dd4hep::mm` directly in the code. It's more readable, follows the DDHep style and the compiler will optimise it out at a compilation time.

**Comment by slomeo** [comment]
@cvuosalo : do I need to continue this PR and then pass to check, by an Analyzer, the Muon Hits (as you asked, putting @civanch in cc) or not?

**Comment by cvuosalo** [comment]
A function like `convertDD4hepUnitsToMm` is simple, self-documenting, and it hides implementation details. On the other hand, I have always found syntax like `length / dd4hep::mm` confusing. This syntax is the inverse of proper dimensional analysis, where one would write, for example, "5 cm * (10 mm / cm) = 50 mm". We have clear syntax in our XML files, e.g., `value="608*mm"`, where we multiply by mm to get the value in mm. But `length / dd4hep::mm` seems to say one divides by mm to get mm, which is the inverse of the XML syntax. I find it confusing, and I bet some new CMSSW developers might get confused by it.

The units conventions in CMSSW are already confusing enough, as shown by previous comments in this thread. I think it is worthwhile to promote extra clarity by spelling out in words what we are doing when changing units. I would hope that the compiler can optimize a simple template function like  `convertDD4hepUnitsToMm` to minimize any performance cost.

**Comment by civanch** [comment]
@cvuosalo , transformation  "length / dd4hep::mm" is fine to me, because we are switching from one system (DD4Hep) to another (Geant4). We do this everywhere in Geant4 when make a switch. In CMSSW there is extra complication, because Geant4 units are used in SIM sub-libraries, the rest using CMSSW units. Whatever you do, there is a room for misunderstanding.

**Comment by cvuosalo** [comment]
@slomeo The priority is to check muon hits. After that is done and any bugs related to missing muon hits are fixed, then you can come back to this PR, which is low priority.

**Comment by slomeo** [comment]
> @slomeo The priority is to check muon hits. After that is done and any bugs related to missing muon hits are fixed, then you can come back to this PR, which is low priority.

@cvuosalo : ok

**Comment by slomeo** [comment]
@cvuosalo : the Muon Sim Hits are validated, so I'll re-start with this PR

**Comment by slomeo** [comment]
@cvuosalo : just an example,

DDD: (1) detid: 637566989 name: MB1RPC_IGasLeft number of Strips: 90
 (2) dpar.size() == 3, width: 1030 length: 584.5 thickness: 2
 (4), tran.x() 4136.75 tran.y(): 376 tran.z(): -4704.5

DD4hep:
(1), detid: 637566989 name: MB1RPC_IGasLeft number of Strips: 90
 (2), dd4hep::Box, width: 1030 length: 584.5 thickness: 2
 (4), tran.x(): 4136.75 tran.y(): 376 tran.z(): -4704.5

**Comment by slomeo** [comment]
@cvuosalo @bsunanda @civanch @ianna : I'm working in  CMSSW_11_3_X_2020-12-02-2300 and for DDD and DD4HEP, when I perform a "cmsRun CondTools/Geometry/test/rpcgeometrywriter.py " I find that myLog.log and myfile.db are "fine" but I see this message:

Fatal system signal has occurred during exit
Aborted

This message was not present in the last CMSSW_11_2 IB. Why?

If it is possible, please perform a "please test" in order to check this PR

**Comment by makortel** [comment]
> I find that myLog.log and myfile.db are "fine" but I see this message:
> 
> Fatal system signal has occurred during exit
> Aborted

We have an open issue on that error #32045 (basically something goes badly wrong during the exit of cmsRun). Do you see any other error messages? Could you try to run with `gdb` and provide the stack trace of the crash point? Or give a recipe to reproduce?

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by slomeo** [comment]
> We have an open issue on that error #32045 (basically something goes badly wrong during the exit of cmsRun). Do you >see any other error messages? Could you try to run with `gdb` and provide the stack trace of the crash point? Or give a >recipe to reproduce?

@makortel , this the recipe:
1) In a new IB, after a "cmsenv", please perform "git cms-rebase-topic slomeo:MigrationToDD4hepForRPCdbLoader"
2) scram b -j 10
3) cmsRun CondTools/Geometry/test/rpcgeometrywriter.py (and you can see the error for DDD Geometry)
4) edit "CondTools/Geometry/test/rpcgeometrywriter.py" setting "True" instead "False" 
5) rm myfile.db
6)  cmsRun CondTools/Geometry/test/rpcgeometrywriter.py (and you can see the error for DD4HEP Geometry)

**Comment by cvuosalo** [comment]
+1

**Comment by makortel** [comment]
@slomeo Thanks, I was able reproduce and got a stack trace in gdb (with `-g -O0`) 
```
Thread 1 "cmsRun" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) where
#0  0x0000000000000000 in ?? ()
#1  0x00007fffed12b7f1 in coral::MessageStream::doOutput() ()
   from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-03-1100/external/slc7_amd64_gcc900/lib/liblcg_CoralBase.so
#2  0x00007fffd6c5a4a7 in coral::ConnectionService::ConnectionPool::~ConnectionPool() ()
   from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-03-1100/external/slc7_amd64_gcc900/lib/liblcg_ConnectionService.so
#3  0x00007fffd6c5a6e9 in coral::ConnectionService::ConnectionPool::~ConnectionPool() ()
   from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-03-1100/external/slc7_amd64_gcc900/lib/liblcg_ConnectionService.so
#4  0x00007fffd6c61c51 in coral::ConnectionService::ConnectionService::~ConnectionService() ()
   from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-03-1100/external/slc7_amd64_gcc900/lib/liblcg_ConnectionService.so
#5  0x00007fffd6c61ce9 in coral::ConnectionService::ConnectionService::~ConnectionService() ()
   from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-03-1100/external/slc7_amd64_gcc900/lib/liblcg_ConnectionService.so
#6  0x00007fffed159643 in coral::Context::~Context() ()
   from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-03-1100/external/slc7_amd64_gcc900/lib/liblcg_CoralKernel.so
#7  0x00007ffff503fce9 in __run_exit_handlers () from /lib64/libc.so.6
#8  0x00007ffff503fd37 in exit () from /lib64/libc.so.6
#9  0x00007ffff502855c in __libc_start_main () from /lib64/libc.so.6
#10 0x000000000040f7f8 in _start ()
```
(certainly out of scope of this PR)

**Comment by slomeo** [comment]
@cvuosalo : while I'm waiting for this PR will be approved and merged I'll update the Ideal Reco DB migration for CSC in order to check if "mm" is the default unit.

**Comment by slomeo** [comment]
Hi all, do you agree to approve this PR? or do I have to wait till #32045 will be merged?

**Comment by silviodonato** [comment]
@ggovi could you have a look?

**Comment by cvuosalo** [comment]
@slomeo I don't think there is any need to wait for #32045. It is unrelated.

**Comment by ggovi** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**